### PR TITLE
Formalize grammar

### PR DESCRIPTION
**Pull request by frostburn**
Replace the parser with a PEG.js AST parser and an AST evaluator.

ref #21

### CODE REVIEW

**Comment by frostburn** [comment]
Tests pass over at `scale-workshop` too if this branch is used as the dependency.

File: src/sw2.pegjs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+// Separator, Space
+Zs = [\u0020\u00A0\u1680\u2000-\u200A\u202F\u205F\u3000]
+
+LineTerminator
+  = [\n\r\u2028\u2029]
+
+_ = Whitespace*
+
+Expression
+  = head:Term tail:(_ ('+' / '-') _ Term)* {
```
Comment (frostburn) [inline src/sw2.pegjs:100]: `head:Term tail:(_ @('+' / '-') _ @Term)*`

**Comment by frostburn** [comment]
Most of these numbers should be `int = num:$[0-9]+ { return parseInt(num, 10) }`.

**Comment by frostburn** [comment]
Ran this by some scales from the Xenharmonic Alliance Discord. Seems to produce the same results.

**Comment by arseniiv** [comment]
I see signed integers and comma-decimals should be accepted by the parser; it’s probably hard to check if the whole expression represents something well-defined at parse time but maybe this wasn’t intended.

**Comment by frostburn** [comment]
> I see signed ~integers~ fractions and comma-decimals should be accepted by the parser; it’s probably hard to check if the whole expression represents something well-defined at parse time but maybe this wasn’t intended.
> 
> SW2 doesn’t accept the scale or/and ignores those entries. Even in SW2, `-3/2` and `-1,5` might have been understood as `2/3` as this additive arithmetic is multiplicative on ratios, but I get why it could’ve been overlooked or intentionally left undefined, as it’s doubly confusing.

This is a re-implementation of the existing and poorly laid out grammar of Scale Workshop 2. I regret allowing `-3/2` to mean `2/3`, but that ship has sailed. I will do better in SW3 if the community allows it.

**Comment by arseniiv** [comment]
> I regret allowing -3/2 to mean 2/3,

But it doesn’t (by itself as a whole row). Then it’s good. But the parser as it is allows it, so transitioning to SW3 parser should catch these places.

**Comment by frostburn** [comment]
> > I regret allowing -3/2 to mean 2/3,
> 
> But it doesn’t (by itself as a whole row). Then it’s good. But the parser as it is allows it, so transitioning to SW3 parser should catch these places.

That is simply due to a fortunate bug. It was always meant to parse and has been fixed in the latest deployment.

**Pull request merged**